guppylang/definition/custom.py

```diff
@@ -138,12 +144,7 @@ def load(
                 "This function does not support usage in a higher-order context",
                 node,
             )
-
-        if self.ty.parametrized:
-            raise InternalGuppyError(
-                "Can't yet generate higher-order versions of custom functions. This "
-                "requires generic function *definitions*"
-            )
+        assert len(self.ty.params) == len(type_args)
```

Comment:
There are non-polymorphic custum Guppy functions whose type cannot be represented in the type system. One example are functions that are implemented via dunder methods (e.g. `iter(x)`, `len(x)`, ...). Therefore, these need `self.higher_order_value=True` which cannot be inferred from the type.

Reply:
> Does RawCustomFunctionDef actually need compile_call or could that be moved into CustomFuncDef?

This is a hack to allow stuff like

https://github.com/CQCL/guppylang/blob/11a1619ef75da82584cd7bddb467e8c19a21aae8/guppylang/prelude/_internal.py#L257-L274

since `Int.__float__` etc are `RawCustomFunctionDef`s. I agree this is not nice but hopefully only a temporary hack. Once we have proper linking, we can define `truediv` etc via actual Guppy function defs and no longer need these `CustomCallCompiler`s.